managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/SharedEntryCacheImpl.java

```diff
@@ -0,0 +1,223 @@
+    }
+
+    @Override
+    public String getName() {
+        return ml.getName();
+    }
+
+    @Override
+    public boolean insert(EntryImpl entry) {
+        return entryCacheManager.insert(entry);
+    }
+
+    @Override
+    public void invalidateEntries(PositionImpl lastPosition) {
+        // No-Op. The cache invalidation is based only on rotating the segment buffers
+    }
+
+    @Override
+    public void invalidateEntriesBeforeTimestamp(long timestamp) {
+        // No-Op. The cache invalidation is based only on rotating the segment buffers
+    }
+
+    @Override
+    public void invalidateAllEntries(long ledgerId) {
+        // No-Op. The cache invalidation is based only on rotating the segment buffers
+    }
+
+    @Override
+    public void clear() {
+        // No-Op. The cache invalidation is based only on rotating the segment buffers
+    }
+
+    private static final Pair<Integer, Long> NO_EVICTION = Pair.of(0, 0L);
+
+    @Override
+    public Pair<Integer, Long> evictEntries(long sizeToFree) {
+        return NO_EVICTION;
+    }
+
+    @Override
+    public void asyncReadEntry(ReadHandle lh, long firstEntry, long lastEntry, boolean isSlowestReader,
+                               AsyncCallbacks.ReadEntriesCallback callback, Object ctx) {
+        final long ledgerId = lh.getId();
+        final int entriesToRead = (int) (lastEntry - firstEntry) + 1;
+
+        if (log.isDebugEnabled()) {
+            log.debug("[{}] Reading entries range ledger {}: {} to {}", ml.getName(), ledgerId, firstEntry, lastEntry);
+        }
+
+        List<Entry> cachedEntries = new ArrayList<>(entriesToRead);
+        long totalCachedSize = entryCacheManager.getRange(ledgerId, firstEntry, lastEntry, cachedEntries);
+
+        if (cachedEntries.size() == entriesToRead) {
+            final List<Entry> entriesToReturn = Lists.newArrayListWithExpectedSize(entriesToRead);
+            // All entries found in cache
+            for (Entry entry : cachedEntries) {
+                entriesToReturn.add(EntryImpl.create((EntryImpl) entry));
+                entry.release();
+            }
+            // All entries found in cache
+            entryCacheManager.getFactoryMBean().recordCacheHits(entriesToReturn.size(), totalCachedSize);
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Ledger {} -- Found in cache entries: {}-{}", ml.getName(), ledgerId, firstEntry,
+                        lastEntry);
+            }
+            callback.readEntriesComplete(entriesToReturn, ctx);
+
+        } else {
+            if (!cachedEntries.isEmpty()) {
+                cachedEntries.forEach(entry -> entry.release());
+            }
+
+            // Read all the entries from bookkeeper
+            lh.readAsync(firstEntry, lastEntry).thenAcceptAsync(
+                    ledgerEntries -> {
+                        requireNonNull(ml.getName());
+                        requireNonNull(ml.getExecutor());
+
```

Comment:
I guess PendingReadsManager should be adapted and used here? It was introduced by  #17241 and resulted in huge improvements.